### PR TITLE
Issue 2 - Splitting successor generation and applicability check

### PR DESCRIPTION
**Pull request by abcorrea**

### CODE REVIEW

File: src/search/search_engines/breadth_first_search.cc [hunk 1]
```diff
@@ -59,14 +59,13 @@ int BreadthFirstSearch<PackedStateT>::search(const Task &task,
 
         assert(sid.id() >= 0 && (unsigned) sid.id() < space.size());
 
-        auto successors = generator.generate_successors(task.actions, packer.unpack(space.get_state(sid)), task.static_info);
+        const auto &state = packer.unpack(space.get_state(sid));
```
Comment (gfrances) [inline src/search/search_engines/breadth_first_search.cc:62]: This should not be a reference, as we're returning an object in the unpack method!

File: src/search/task.cc [hunk 2]
```diff
@@ -36,22 +36,22 @@ void Task::create_empty_initial_state(size_t number_predicates)
...
 }
 
 void Task::dump_state(DBState s) const
 {
     /*
      * Output initial state in a human readable way.
      */
-    for (size_t j = 0; j < s.nullary_atoms.size(); ++j) {
-        if (s.nullary_atoms[j])
+    auto nullary_atoms = s.get_nullary_atoms();
```
Comment (gfrances) [inline src/search/task.cc:46]: const auto&

File: src/search/task.cc [hunk 3]
```diff
@@ -36,22 +36,22 @@ void Task::create_empty_initial_state(size_t number_predicates)
...
+    for (size_t j = 0; j < nullary_atoms.size(); ++j) {
+        if (nullary_atoms[j])
             cout << predicates[j].getName() << ", ";
     }
-    for (size_t i = 0; i < s.relations.size(); ++i) {
+    auto relations = s.get_relations();
```
Comment (gfrances) [inline src/search/task.cc:51]: I prefer `const auto&` here - honestly don't remember if it makes a difference according to the spec, but I find it clearer...

File: src/search/heuristics/goalcount.cc [hunk 4]
```diff
@@ -10,14 +10,15 @@ int Goalcount::compute_heuristic(const DBState &s, const Task &task) {
...
-                                          s.nullary_atoms);
+                                          nullary_atoms);
     h += compute_reached_nullary_atoms(task.goal.negative_nullary_goals,
-                                       s.nullary_atoms);;
+                                       nullary_atoms);;
 
     for (const AtomicGoal &atomicGoal : task.goal.goal) {
         assert(atomicGoal.predicate==
-            s.relations[atomicGoal.predicate].predicate_symbol);
+            s.get_relations()[atomicGoal.predicate].predicate_symbol);
```
[the remark this answers is not in the document]
Comment (abcorrea) [reply src/search/heuristics/goalcount.cc:21]: This can be done but I would do this outside this issue.

**Pull request merged**